v2/jdbc-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/JdbcToBigQuery.java

```diff
@@ -175,12 +202,71 @@ static PipelineResult run(JdbcToBigQueryOptions options, Write<TableRow> writeTo
     /*
      * Step 2: Append TableRow to an existing BigQuery table
      */
-    rows.apply("Write to BigQuery", writeToBQ);
+    WriteResult writeResult = rows.apply("Write to BigQuery", writeToBQ);
+
+    /*
+     * Step 3.
+     * If using Storage Write API, capture failed inserts and either
+     *   a) write error rows to DLQ
+     *   b) fail the pipeline
+     */
+    if (options.getUseStorageWriteApi() || options.getUseStorageWriteApiAtLeastOnce()) {
+      PCollection<BigQueryInsertError> insertErrors =
+          BigQueryIOUtils.writeResultToBigQueryInsertErrors(writeResult, options);
+
+      if (!Strings.isNullOrEmpty(options.getOutputDeadletterTable())) {
+        /*
+         * Step 3a.
+         * Elements that failed inserts into BigQuery are extracted and converted to FailsafeElement
+         */
+        PCollection<FailsafeElement<String, String>> failedInserts =
+            insertErrors
+                .apply(
+                    "WrapInsertionErrors",
+                    MapElements.into(FAILSAFE_ELEMENT_CODER.getEncodedTypeDescriptor())
+                        .via((BigQueryInsertError e) -> wrapBigQueryInsertError(e)))
+                .setCoder(FAILSAFE_ELEMENT_CODER);
+
+        /*
+         * Step 3a Contd.
+         * Insert records that failed insert into deadletter table
+         */
+        failedInserts.apply(
+            "WriteFailedRecords",
+            ErrorConverters.WriteStringMessageErrors.newBuilder()
+                .setErrorRecordsTable(options.getOutputDeadletterTable())
+                .setErrorRecordsTableSchema(ResourceUtils.getDeadletterTableSchemaJson())
+                .setUseWindowedTimestamp(false)
+                .build());
+      } else {
+        /*
+         * Step 3b.
+         * Fail pipeline upon write errors if no DLQ was specified
+         */
+        insertErrors.apply(ParDo.of(new ThrowWriteErrorsDoFn()));
```

Comment:
iiuc now error will be thrown if there is a failed write record and dlq table is not specified (the default), which means some pipeline succeeded before now would fail. This is a breaking change. However I understand previous behavior (silently drop record) was bad either. Shall we call this out somewhere, while still give user an option preserve the old behavior if they notices breakage, that is to drop bad records like the previous behavior (e.g. set deadletter table to a magic name like "ignored"

Reply:
I don't think we should give an option to ignore this - it is silent data loss, so I think we should force users to apply it. Agreed on calling it out somewhere - the PR description will get auto-added to the release notes, so maybe we could make that more descriptive? Alternately, we could coordinate with the next release manager to make sure it gets mentioned explicitly

Reply:
Thanks, the current form sounds good to me